product_template_multi_link/models/product_template.py

```diff
@@ -103,3 +104,37 @@ def get_links_by_code(self, code):
         return self.product_template_link_ids.filtered(
             lambda r: r.type_id.code == code and r.is_link_active
         )
+
+    @api.model
+    def _name_search(
+        self, name, args=None, operator="ilike", limit=100, name_get_uid=None
+    ):
+        # NOTE: Odoo limits the search on the name of templates as soon as
+        # the 'id' field is present in 'args' domain, and this 'id' criteria is
+        # set by the view on purpose to avoid a search on variants.
+        # Improve this search by looking also on template's default_code
+        # if there is only a domain on 'id'.
+        search_default_code = self.env.context.get("name_search_default_code")
+        if name and len(args or []) == 1 and args[0][0] == "id" and search_default_code:
+            args = expression.AND(
+                [
+                    args,
+                    expression.OR(
+                        [
+                            [("default_code", operator, name)],
+                            [(self._rec_name, operator, name)],
+                        ]
+                    ),
+                ]
+            )
+            # Reset 'name' so base '_name_search' won't add '_rec_name'
+            # to 'args' (already added above).
+            # See 'odoo.models.BaseModel._name_search'.
+            name = ""
```

Comment:
Actually you could simplify your code by returning the result of `_search` in your function as it's what Odoo does in lots of places when a `name_search` has to consider multiple fields.

Something along these lines could work just as well:
```
    @api.model
    def _name_search(self, name, args=None, operator='ilike', limit=100, name_get_uid=None):
        search_default_code = self.env.context.get("name_search_default_code")
        if search_default_code and not (operator == 'ilike' and not (name or '').strip()):
            args = args or []
            domain = ['|', ('name', operator, name), (search_default_code, operator, name)]
            return self._search(expression.OR([domain, args]), limit=limit, access_rights_uid=name_get_uid)
        return super()._name_search(name, args=args, operator=operator, limit=limit, name_get_uid=name_get_uid)
```
at least it's easier from v16.0, cf: https://github.com/odoo/odoo/pull/86588